lib/uint.ml

```diff
@@ -235,6 +233,15 @@ module Uint128 = struct
     else
       Uint64.zero, Uint64.shift_right_logical (fst a) (by-64)
 
+  let shift_right a by =
+    let upper_bit = shift_left one (num_bits - 1) in
+    let has_msb = logand a upper_bit = upper_bit in
+    let rest = shift_right_logical a by in
+    if has_msb then
+      logor rest upper_bit
+    else
+      rest
```

Comment:
Hi, if I remember correctly, arithmetic shift right replicates the sign bit. It looks as if the above code copies this most significant bit into the result without replication.

Reply:
Oops, yes it does; good catch. Done.